### PR TITLE
Allow ErrorDecoder.decode(...) to return null which signals that no ...

### PR DESCRIPTION
**Pull request by rfalke**
... exception should be thrown. This allows to react to non-200 statuses not only with an
exception but also to return it in the return object.

### CODE REVIEW

File: core/src/main/java/feign/Logger.java [hunk 1]
```diff
@@ -107,7 +107,7 @@ protected Response logAndRebufferResponse(String configKey, Level logLevel, Resp
     return response;
   }
 
-  protected IOException logIOException(String configKey, Level logLevel, IOException ioe, long elapsedTime) {
+  protected void logIOException(String configKey, Level logLevel, IOException ioe, long elapsedTime) {
```
Comment (spencergibb) [inline core/src/main/java/feign/Logger.java]: Why this change? It breaks an API?

File: core/src/main/java/feign/codec/ErrorDecoder.java [hunk 2]
```diff
@@ -76,13 +73,15 @@
...
-   * application-specific exception decoded by the implementation. If the throwable is retryable, it
-   * should be wrapped, or a subtype of {@link RetryableException}
+   * @return Exception An Exception or <code>null</code>. IOException, if there was a network error
+   * reading the response or an application-specific exception decoded by the implementation. If
+   * the throwable is retryable, it should be wrapped, or a subtype of {@link RetryableException}.
+   * If <code>null</code> is returned no exception is thrown and the normal {@link Decoder} will
+   * be asked to decode the response.
    */
-  public Exception decode(String methodKey, Response response);
+  Exception decode(String methodKey, Response response);
```
Comment (spencergibb) [inline core/src/main/java/feign/codec/ErrorDecoder.java]: Why change visibility?

File: core/src/main/java/feign/codec/ErrorDecoder.java [hunk 3]
```diff
@@ -76,13 +73,15 @@
...
 
-  public static class Default implements ErrorDecoder {
+  class Default implements ErrorDecoder {
```
Comment (spencergibb) [inline core/src/main/java/feign/codec/ErrorDecoder.java]: Why?

File: core/src/main/java/feign/codec/ErrorDecoder.java [hunk 4]
```diff
@@ -108,7 +107,7 @@ public Exception decode(String methodKey, Response response) {
    * Decodes a {@link feign.Util#RETRY_AFTER} header into an absolute date, if possible. <br> See <a
    * href="https://tools.ietf.org/html/rfc2616#section-14.37">Retry-After format</a>
    */
-  static class RetryAfterDecoder {
+  class RetryAfterDecoder {
```
Comment (spencergibb) [inline core/src/main/java/feign/codec/ErrorDecoder.java]: Why?

File: core/src/main/java/feign/SynchronousMethodHandler.java [hunk 5]
```diff
@@ -129,26 +139,31 @@ Object executeAndDecode(RequestTemplate template) throws Throwable {
...
-      throw errorReading(request, response, e);
+      return decodeWithDecoder(response);
     } finally {
       if (shouldClose) {
         ensureClosed(response.body());
       }
     }
   }
 
+  Object decodeWithDecoder(Response response) throws Throwable {
```
Comment (spencergibb) [inline core/src/main/java/feign/SynchronousMethodHandler.java]: Why move this?

File: core/src/main/java/feign/SynchronousMethodHandler.java [hunk 6]
```diff
@@ -105,14 +104,25 @@ Object executeAndDecode(RequestTemplate template) throws Throwable {
...
-    boolean shouldClose = true;
     try {
       if (logLevel != Logger.Level.NONE) {
         response =
-            logger.logAndRebufferResponse(metadata.configKey(), logLevel, response, elapsedTime);
+                logger.logAndRebufferResponse(metadata.configKey(), logLevel, response, elapsedTime);
         // ensure the request is set. TODO: remove in Feign 10
         response.toBuilder().request(request).build();
       }
+      return decodeResponse(response);
```
Comment (spencergibb) [inline core/src/main/java/feign/SynchronousMethodHandler.java]: Moving this to another method just makes it harder to see what you are actually changing.

File: core/src/main/java/feign/Logger.java (hunk 1 above)
Comment (rfalke) [inline core/src/main/java/feign/Logger.java]: 1. The return value is unused. 
1. The method is neither a command nor a query.

File: core/src/main/java/feign/codec/ErrorDecoder.java (hunk 2 above)
Comment (rfalke) [inline core/src/main/java/feign/codec/ErrorDecoder.java]: Interface methods are public by default. The 'public' is redundant here. The IDE flagged this.

I will re-add it if the code style guide requires it.

File: core/src/main/java/feign/codec/ErrorDecoder.java (hunk 4 above)
Comment (rfalke) [inline core/src/main/java/feign/codec/ErrorDecoder.java]: Same.

File: core/src/main/java/feign/SynchronousMethodHandler.java (hunk 6 above)
Comment (rfalke) [inline core/src/main/java/feign/SynchronousMethodHandler.java]: The old method was already big. Too big for my taste. Even the new decodeResponse() is on the upper limit for my taste.

File: core/src/main/java/feign/SynchronousMethodHandler.java (hunk 5 above)
Comment (rfalke) [inline core/src/main/java/feign/SynchronousMethodHandler.java]: So that is nearer to the one-and-only caller.

**Comment by rfalke** [comment]
Just tell me which changes should be reverted and I will do so.

File: core/src/main/java/feign/Logger.java (hunk 1 above)
Comment (adriancole) [inline core/src/main/java/feign/Logger.java]: not good enough reasons, as this would break people who are extending this regardless.

a good enough reason is that this method was exposed protected, but not yet released! (unless I'm mistaken) This means that no one would have been able to extend it yet, so we wouldn't be breaking anyone.

File: core/src/main/java/feign/codec/ErrorDecoder.java (hunk 2 above)
Comment (adriancole) [inline core/src/main/java/feign/codec/ErrorDecoder.java]: it is more about making distracting change in the middle of a request, not about some style guide ruling. please be considerate to the time it takes to figure out what is actually needed or not.

File: core/src/test/java/feign/FeignTest.java [hunk 7]
```diff
@@ -548,7 +549,7 @@ public Object decode(Response response, Type type) throws IOException {
   @Test
   public void decoderCanThrowUnwrappedExceptionInDecode404Mode() throws Exception {
     server.enqueue(new MockResponse().setResponseCode(404));
-    thrown.expect(NoSuchElementException.class);
+    thrown.expectCause(isA(NoSuchElementException.class));
```
Comment (adriancole) [inline core/src/test/java/feign/FeignTest.java]: was this needed?

**Comment by rfalke** [comment]
Is superseded by #487 and #488.
